README.md
# Cropty

A little tool for doing encryption using a combination of RSA and AEP, using the
[cryptonite](https://hackage.haskell.org/package/cryptonite) library for cryptography
and [commander-cli](https://hackage.haskell.org/package/commander-cli) for its CLI
interface.

It is meant for use with very large files. I've tested it on a 6 Gigabyte file and it
works within seconds for all functions.

## Installation

To install, run the following in this directory:

```bash
cabal install --installdir=.
```

## Example Usage

```bash
$ export IDENTITY_FILE=my_identity
$ cropty identity populate -s 512
$ echo "Hello, world! This is my secret message" > message
$ cropty encrypt my_identity.public message.encrypted message
$ cropty decrypt message.encrypted message.decrypted
$ cat message.decrypted

$ cropty sign message.encrypted message.signature
$ cropty verify message.encrypted message.signature my_identity.public
```

```bash
$ cropty help

usage:
name: cropty
|
+- subprogram: help
|
+- description: A CLI for simple cryptographic tasks.
|
`- required env: IDENTITY_FILE :: [Char], the environment variable containing the filename of my private key, with my public key located at $IDENTITY_FILE.public
   |
   +- subprogram: identity
   |  |
   |  `- subprogram: populate
   |     |
   |     +- description: Populate the IDENTITY_FILE location with a new RSA private key, writing the public key to IDENTITY_FILE.public
   |     |
   |     `- option: -s <key-size :: Int>, the size of your RSA key pair (e.g. 1024, 2048, 4096), defaulting to 2048
   |
   +- subprogram: encrypt
   |  |
   |  +- description: Encrypt a file for decryption by someone with the private key matching the public key you pass in.
   |  |
   |  `- argument: public-key-filepath :: [Char], the file where your friend's public key is loaded
   |     |
   |     `- argument: destination-filepath :: [Char], the file your encrypted message will be written to
   |        |
   |        `- argument: plaintext-filepath :: [Char], the file containing the plaintext you want to encrypt
   |
   +- subprogram: decrypt
   |  |
   |  +- description: Decrypt a file encrypted for you with your private key.
   |  |
   |  `- argument: encrypted-filename :: [Char], the file which is currently encrypted
   |     |
   |     `- argument: decrypted-filename :: [Char], the file your plaintext will be written to
   |
   +- subprogram: sign
   |  |
   |  +- description: Sign a file with your private key.
   |  |
   |  `- argument: filename-to-sign :: [Char], the file whose contents you will sign
   |     |
   |     `- argument: signature-filename :: [Char], the file where you will write out the signature
   |
   `- subprogram: verify
      |
      +- description: Verify that a signature belongs to the owner of the private key associated to the one you've passed in
      |
      `- argument: filename-signed :: [Char], the file which was signed
         |
         `- argument: signature-filename :: [Char], the file with the signature
            |
            `- argument: signer-pubkey :: [Char], the file containing the public key of who produced this signature by signing this file
```


CHANGELOG.md
# Revision history for cropty

## 0.1.0.0 -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


test/Test.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE BlockArguments #-}
module Main where

import Cropty
import Control.Monad.IO.Class (liftIO)
import Hedgehog
import Hedgehog.Gen
import Hedgehog.Range
import Control.Monad (guard)

main :: IO ()
main = do
  keypairs <- sequence
    [ (\private -> (private, privateToPublic private)) <$> generatePrivateKey s
    | s <- [256, 512]
    ]
  let
    nTests = 1000
    roundTrip gen = withTests nTests $ property do
      (privateKey, publicKey) <- forAll (element keypairs)
      x <- forAll gen
      msg <- liftIO (encrypt publicKey x)
      y <- liftIO (decrypt privateKey msg)
      x === y
    signAndVerify gen = withTests nTests $ property do
      (privateKey, publicKey) <- forAll (element keypairs)
      x <- forAll gen
      sig <- liftIO (sign privateKey x)
      assert (verify publicKey x sig)
  guard =<< checkParallel (Group "Encryption/Decryption" [
        ("Encrypt/Decrypt UTF-8",
          roundTrip (utf8 (linearFrom 0 1000 10000) unicodeAll)
        )
      , ("Encrypt/Decrypt Bytes",
          roundTrip (bytes (linearFrom 0 1000 10000))
        )
      , ("Encrypt/Decrybt Trailing Zeros",
          roundTrip ((<>) <$> bytes (linearFrom 0 1000 10000) <*> pure "\0\0\0\0\0\0")
        )
      , ("Encrypt/Decrypt Leading Zeros",
          roundTrip ((<>) <$> pure "\0\0\0\0\0\0" <*> bytes (linearFrom 0 1000 10000))
        )
    ])
  guard =<< checkParallel (Group "Signing/Verification" [
        ("Sign/Verify UTF-8",
          signAndVerify (utf8 (linearFrom 0 1000 10000) unicodeAll)
        )
      , ("Sign/Verify Bytes",
          signAndVerify (bytes (linearFrom 0 1000 10000))
        )
      , ("Sign/Verify: Trailing Zeros",
          roundTrip ((<>) <$> bytes (linearFrom 0 1000 10000) <*> pure "\0\0\0\0\0\0")
        )
      , ("Sign/Verify: Leading Zeros",
          roundTrip ((<>) <$> pure "\0\0\0\0\0\0" <*> bytes (linearFrom 0 1000 10000))
        )
    ])
      


src/Cropty.hs
{-# LANGUAGE DeriveAnyClass #-}
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE NamedFieldPuns #-}
{-# LANGUAGE LambdaCase #-}
module Cropty where

import Data.ByteString (ByteString)
import GHC.Generics (Generic)
import Data.Binary (Binary)
import Control.Monad (guard)
import Control.Exception (SomeException, catch, Exception, throwIO)
import qualified Crypto.Cipher.AES as AES
import qualified Crypto.Cipher.Types as Cipher
import qualified Crypto.Data.Padding as Padding
import qualified Crypto.Error as Error
import qualified Crypto.Hash.Algorithms as Hash
import qualified Crypto.PubKey.RSA as RSA
import qualified Crypto.PubKey.RSA.OAEP as RSA.OAEP
import qualified Crypto.PubKey.RSA.PSS as RSA.PSS
import qualified Crypto.Random as Random
import qualified Data.ByteString as ByteString
import qualified Data.ByteString.Char8 as ByteString.Char8

newtype PrivateKey = PrivateKey
  { privateKey :: RSA.PrivateKey }
  deriving (Show, Read, Eq)

data PublicKey = PublicKey
  { publicKey :: RSA.PublicKey }
  deriving (Show, Read, Eq)

privateToPublic :: PrivateKey -> PublicKey
privateToPublic = PublicKey . RSA.private_pub . privateKey

data GenerationException = GenerationException String
  deriving Show

instance Exception GenerationException

generatePrivateKey :: Int -> IO PrivateKey
generatePrivateKey n =
    if n >= 256
      then (PrivateKey . snd) <$> RSA.generate n 65537
      else throwIO (GenerationException "size of key must be greater than or equal to 256")

encryptSmall :: PublicKey -> ByteString -> IO (Either RSA.Error ByteString)
encryptSmall (PublicKey pub) message =
    RSA.OAEP.encrypt (RSA.OAEP.defaultOAEPParams Hash.SHA512) pub message

decryptSmall :: PrivateKey -> ByteString -> IO (Either RSA.Error ByteString)
decryptSmall (PrivateKey priv) message =
    RSA.OAEP.decryptSafer (RSA.OAEP.defaultOAEPParams Hash.SHA512) priv message

newtype Key =
    Key { keyBytes :: ByteString }

generateKey :: IO Key
generateKey =
    Key <$> Random.getRandomBytes 32

data Message = Message
  { encryptedKey :: ByteString
  , encryptedBytes :: ByteString
  } deriving (Show, Read, Generic, Binary)

data EncryptionException = EncryptionException String
  deriving Show

instance Exception EncryptionException

encrypt :: PublicKey -> ByteString -> IO Message
encrypt publicKey message = do
  key <- generateKey
  encryptSmall publicKey (keyBytes key) >>= \case
    Left rsaError -> throwIO $ EncryptionException (show rsaError)
    Right encryptedKey -> case Cipher.cipherInit (keyBytes key) of
      Error.CryptoFailed e -> throwIO $ EncryptionException (show e)
      Error.CryptoPassed (c :: AES.AES256) -> do
        let encryptedBytes = Cipher.ecbEncrypt c paddedMessage
        pure Message{encryptedKey, encryptedBytes}
  where
    paddingSize =
      16 - (ByteString.length message + 1) `mod` 16
    paddedMessage =
      ByteString.concat
        [ ByteString.singleton (fromIntegral paddingSize)
        , ByteString.replicate paddingSize 0
        , message
        ]

data DecryptionException = DecryptionException String
  deriving Show

instance Exception DecryptionException

decrypt :: PrivateKey -> Message -> IO ByteString
decrypt privateKey Message{encryptedKey, encryptedBytes} = do
  decryptSmall privateKey encryptedKey >>= \case
    Left rsaError -> throwIO $ DecryptionException (show rsaError)
    Right decryptedKey -> case Cipher.cipherInit decryptedKey of
      Error.CryptoFailed e -> throwIO $ DecryptionException (show e)
      Error.CryptoPassed (c :: AES.AES256) -> do
        let decryptedBytes = Cipher.ecbDecrypt c encryptedBytes
        if ByteString.length decryptedBytes > 0 then do
          let paddingSize = fromIntegral (ByteString.index decryptedBytes 0)
          pure $ snd (ByteString.splitAt (paddingSize + 1) decryptedBytes)
        else throwIO (DecryptionException "Not encrypted by Cropty")

data SignatureException = SignatureException String
  deriving Show

instance Exception SignatureException

sign :: PrivateKey -> ByteString -> IO ByteString
sign (PrivateKey privateKey) bs =
    RSA.PSS.signSafer
      (RSA.PSS.defaultPSSParams Hash.SHA512)
      privateKey
      bs
    >>= either (throwIO . SignatureException . show) pure

verify :: PublicKey -> ByteString -> ByteString -> Bool
verify (PublicKey pubKey) bs sig =
    RSA.PSS.verify (RSA.PSS.defaultPSSParams Hash.SHA512) pubKey bs sig


src/App.hs
{-# LANGUAGE ViewPatterns #-}
{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE TypeOperators #-}
module App (main) where

import Options.Commander
import System.Directory (doesFileExist)
import Cropty
import UnliftIO.IO.File
import System.IO (hPutStrLn, stderr)
import qualified Data.ByteString.Char8 as ByteString.Char8
import qualified Data.ByteString.Lazy as ByteString.Lazy
import qualified Data.ByteString as ByteString
import qualified Data.Binary as Binary

main :: IO ()
main = command_ $ toplevel @"cropty" program
  where
    program = description @"A CLI for simple cryptographic tasks." $
      annotated @"the environment variable containing the filename of my private key, with my public key located at $IDENTITY_FILE.public" $
      env @"IDENTITY_FILE" $ \identityFile ->
      sub @"identity"
        ( sub @"populate"
        $ description @"Populate the IDENTITY_FILE location with a new RSA private key, writing the public key to IDENTITY_FILE.public"
        $ annotated @"the size of your RSA key pair (e.g. 1024, 2048, 4096), defaulting to 2048"
        $ opt @"s" @"key-size" $ \keySize -> raw (populateIdentity keySize identityFile)
        )
      <+> sub @"encrypt"
        ( description @"Encrypt a file for decryption by someone with the private key matching the public key you pass in."
        $ annotated @"the file where your friend's public key is loaded" $ arg @"public-key-filepath" $ \to ->
          annotated @"the file your encrypted message will be written to" $ arg @"destination-filepath" $ \dest ->
          annotated @"the file containing the plaintext you want to encrypt" $ arg @"plaintext-filepath" (raw . encryptFile dest to))
      <+> sub @"decrypt"
        ( description @"Decrypt a file encrypted for you with your private key."
        $ annotated @"the file which is currently encrypted" $ arg @"encrypted-filename" $ \encFilepath ->
          annotated @"the file your plaintext will be written to" $ arg @"decrypted-filename" (raw . decryptFile encFilepath identityFile)
        )
      <+> sub @"sign"
        ( description @"Sign a file with your private key."
        $ annotated @"the file whose contents you will sign" $ arg @"filename-to-sign" $ \fileToSign ->
          annotated @"the file where you will write out the signature" $ arg @"signature-filename" (raw . signFile identityFile fileToSign)
        )
      <+> sub @"verify"
        ( description @"Verify that a signature belongs to the owner of the private key associated to the one you've passed in"
        $ annotated @"the file which was signed" $ arg @"filename-signed" $ \fileSigned ->
          annotated @"the file with the signature" $ arg @"signature-filename" $ \signatureFilename ->
          annotated @"the file containing the public key of who produced this signature by signing this file" $ arg @"signer-pubkey" (raw . verifySignature fileSigned signatureFilename)
        )
    populateIdentity :: Maybe Int -> FilePath -> IO ()
    populateIdentity (maybe 2048 id -> n) identityFilepath = do
      doesFileExist identityFilepath >>= \case
        True -> putStrLn "You already have a populated identity file. Delete it manually if you would like to generate a new one."
        False -> do
          priv <- generatePrivateKey n
          writeBinaryFileDurableAtomic identityFilepath (ByteString.Char8.pack $ show priv)
          writeBinaryFileDurableAtomic (identityFilepath <> ".public") (ByteString.Char8.pack $ show (privateToPublic priv))
    encryptFile :: FilePath -> FilePath -> FilePath -> IO ()
    encryptFile destFilepath toFilePath msgFilepath = do
      pub <- read <$> readFile toFilePath
      secretMsg <- ByteString.readFile msgFilepath
      msg <- encrypt pub secretMsg
      writeBinaryFileDurableAtomic destFilepath (ByteString.Lazy.toStrict $ Binary.encode msg)
    decryptFile :: FilePath -> FilePath -> FilePath -> IO ()
    decryptFile msgFilepath identityFilepath destFilepath = do
      priv <- read <$> readFile identityFilepath
      msg <-  Binary.decode <$> ByteString.Lazy.readFile msgFilepath
      decrypted <- decrypt priv msg
      writeBinaryFileDurableAtomic destFilepath decrypted
    signFile :: FilePath -> FilePath -> FilePath -> IO ()
    signFile identityFilepath encFilepath signatureFilepath = do
      priv <- read <$> readFile identityFilepath
      msg <- ByteString.readFile encFilepath
      signature <- sign priv msg
      writeBinaryFileDurableAtomic signatureFilepath signature
    verifySignature :: FilePath -> FilePath -> FilePath -> IO ()
    verifySignature fileSigned signatureFilename signerFilename = do
      pub <- read <$> readFile signerFilename
      signature <- ByteString.readFile signatureFilename
      msg <- ByteString.readFile fileSigned
      if verify pub msg signature then
        putStrLn "Signature valid."
      else
        putStrLn "Signature invalid."
